### PR TITLE
GH-117442: Check eval-breaker at start (rather than end) of tier 2 loops

### PR DESCRIPTION
**Pull request by markshannon**
Insert a `_TIER2_RESUME_CHECK` at the start of all traces starting at a back edge.
`_JUMP_TO_TOP` no longer performs an eval breaker check.

Also, simplify the handling of the eval breaker in tier 2:
* Always drop to tier 1 if the eval breaker is set.
* Guarantee progress by never entering tier 2 from tier 1 with the eval breaker set.

* Issue: gh-117442

### CODE REVIEW

**Comment by brandtbucher** [comment]
Can you remove the `CHECK_EVAL_BREAKER();` from the `_JUMP_TO_TOP` special-case in `Tools/jit/template.c`?

File: Python/optimizer.c [hunk 1]
```diff
@@ -627,6 +627,9 @@ translate_bytecode_to_trace(
             if (opcode == JUMP_BACKWARD || opcode == JUMP_BACKWARD_NO_INTERRUPT) {
                 instr += 1 + _PyOpcode_Caches[opcode] - (int32_t)oparg;
                 initial_instr = instr;
+                if (opcode == JUMP_BACKWARD) {
+                    ADD_TO_TRACE(_TIER2_RESUME_CHECK, 0, 0, target);
```
Comment (gvanrossum) [inline Python/optimizer.c:631]: Is this accounted for by the RESERVE for this opcode?

**Review by gvanrossum — COMMENTED** [review]
Simple enough, I have one question.

File: Python/optimizer.c (hunk 1 above)
Comment (markshannon) [reply Python/optimizer.c:631]: ~Not sure. I'll add an extra reserve to be on the same side~
No, but this is the first instruction of the trace, so there's no shortage of space.

**Review by brandtbucher — APPROVED** [review]

**Comment by brandtbucher** [comment]
Nice, that ended up being a lot cleaner than I thought it would be!

**Review by gvanrossum — APPROVED** [review]

**Pull request merged**